File: package/typescript/src/countConnections.test.ts
import { countConnections, Log, CountConnectionsParam } from './countConnections';

describe('countConnections', () => {
  test('basic test case', () => {
    /**
     * 基本的なテストケース。
     * 時刻ごとの接続数をテストする。
     * endTime=5, period=1 で複数のログを使用して接続数を集計。
     * 結果として [3, 5, 5, 5, 8, 6] が期待される。
     */
    const param: CountConnectionsParam = {
      endTime: 5,
      period: 1,
      logs: [
        { time: 0, nConnect: 3, nDisconnect: 0 },
        { time: 1, nConnect: 2, nDisconnect: 0 },
        { time: 4, nConnect: 5, nDisconnect: 2 },
        { time: 5, nConnect: 3, nDisconnect: 5 },
      ],
    };
    expect(countConnections(param)).toEqual([3, 5, 5, 5, 8, 6]);
  });

  test('no logs', () => {
    /**
     * ログが存在しない場合のテスト。
     * すべての接続数が 0 であることを確認。
     */
    const param: CountConnectionsParam = { endTime: 3, period: 1, logs: [] };
    expect(countConnections(param)).toEqual([0, 0, 0, 0]);
  });

  test('gap logs', () => {
    /**
     * ログの間隔にギャップがある場合のテスト。
     * endTime=6, period=2 で複数のログを使用して接続数を集計。
     * 結果として [0, 4, 4, 5] が期待される。
     */
    const param: CountConnectionsParam = {
      endTime: 6,
      period: 2,
      logs: [
        { time: 1, nConnect: 4, nDisconnect: 0 },
        { time: 3, nConnect: 1, nDisconnect: 1 },
        { time: 6, nConnect: 3, nDisconnect: 2 },
      ],
    };
    expect(countConnections(param)).toEqual([0, 4, 4, 5]);
  });

  test('single log', () => {
    /**
     * 単一のログが存在する場合のテスト。
     * endTime=5, period=1 で1つのログのみ使用し、接続数を集計。
     * 結果として [5, 5, 5, 5, 5, 5] が期待される。
     */
    const param: CountConnectionsParam = {
      endTime: 5,
      period: 1,
      logs: [
        { time: 0, nConnect: 5, nDisconnect: 0 },
      ],
    };
    expect(countConnections(param)).toEqual([5, 5, 5, 5, 5, 5]);
  });

  test('all disconnect', () => {
    /**
     * すべての接続が切断される場合のテスト。
     * 時刻 2 で全ての接続が切断され、結果として [10, 10, 0, 0, 0] が期待される。
     */
    const param: CountConnectionsParam = {
      endTime: 4,
      period: 1,
      logs: [
        { time: 0, nConnect: 10, nDisconnect: 0 },
        { time: 2, nConnect: 0, nDisconnect: 10 },
      ],
    };
    expect(countConnections(param)).toEqual([10, 10, 0, 0, 0]);
  });

  test('partial disconnect', () => {
    /**
     * 部分的に切断が行われる場合のテスト。
     * endTime=4, period=1 で部分的に接続が切断される状況をテスト。
     * 結果として [10, 10, 5, 5, 5] が期待される。
     */
    const param: CountConnectionsParam = {
      endTime: 4,
      period: 1,
      logs: [
        { time: 0, nConnect: 10, nDisconnect: 0 },
        { time: 2, nConnect: 0, nDisconnect: 5 },
      ],
    };
    expect(countConnections(param)).toEqual([10, 10, 5, 5, 5]);
  });

  test('large connect and disconnect', () => {
    /**
     * 非常に多くの接続が行われ、その後すべて切断されるケースのテスト。
     * endTime=100, period=10 で一気に大量の接続があり、その後すべて切断される状況をテスト。
     * 最終的にすべての接続が切断され、結果として [0, 10**10, 10**10, 10**10, 0, 0, 0, 0, 0, 0, 0] が期待される。
     */
    const param: CountConnectionsParam = {
      endTime: 100,
      period: 10,
      logs: [
        { time: 10, nConnect: 10 ** 10, nDisconnect: 0 },
        { time: 40, nConnect: 0, nDisconnect: 10 ** 10 },
      ],
    };
    expect(countConnections(param)).toEqual([0, 10 ** 10, 10 ** 10, 10 ** 10, 0, 0, 0, 0, 0, 0, 0]);
  });

  test('output 1000 entries', () => {
    /**
     * 上限のテスト。
     * endTime=1000, period=1 で各時刻ごとに出力が行われるケースをテスト。
     * すべての時間に対する接続数が正しく計算されているかを確認。
     */
    const logs: Log[] = [];
    for (let i = 0; i <= 1000; i++) {
      logs.push({ time: i, nConnect: 1, nDisconnect: 0 });
    }
    const param: CountConnectionsParam = { endTime: 1000, period: 1, logs };
    const expected = [];
    for (let i = 0; i <= 1000; i++) {
      expected.push(i + 1);
    }
    expect(countConnections(param)).toEqual(expected);
  });

  test('boundary period equals end time', () => {
    /**
     * period が end_time と等しい場合のテスト。
     * endTime=10, period=10 の場合、結果として [10, 10] が期待される。
     */
    const param: CountConnectionsParam = {
      endTime: 10,
      period: 10,
      logs: [
        { time: 0, nConnect: 10, nDisconnect: 0 },
      ],
    };
    expect(countConnections(param)).toEqual([10, 10]);
  });

  test('overlapping connections and disconnections', () => {
    /**
     * 接続と切断が重複して発生する場合のテスト。
     * 同じ時刻で接続と切断が同時に発生する状況をテスト。
     */
    const param: CountConnectionsParam = {
      endTime: 5,
      period: 1,
      logs: [
        { time: 0, nConnect: 5, nDisconnect: 0 },
        { time: 2, nConnect: 3, nDisconnect: 2 }, // 時刻2で3接続、2切断
        { time: 4, nConnect: 0, nDisconnect: 6 }, // 時刻4で全て切断
      ],
    };
    expect(countConnections(param)).toEqual([5, 5, 6, 6, 0, 0]);
  });
});

File: package/go/tests/count_connections_test.go
package tests

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"program-knock/src"
)

func TestCountConnections(t *testing.T) {
	t.Run("Basic", func(t *testing.T) {
		// Basic test case
		// Count connections at each time interval
		// end_time=5, period=1 with multiple logs to aggregate connection counts
		// Expected result: [3, 5, 5, 5, 8, 6]
		param := src.CountConnectionsParam{
			EndTime: 5,
			Period:  1,
			Logs: []src.Log{
				{Time: 0, NConnect: 3, NDisconnect: 0},
				{Time: 1, NConnect: 2, NDisconnect: 0},
				{Time: 4, NConnect: 5, NDisconnect: 2},
				{Time: 5, NConnect: 3, NDisconnect: 5},
			},
		}
		result := src.CountConnections(param)
		expected := []int{3, 5, 5, 5, 8, 6}
		assert.Equal(t, expected, result)
	})

	t.Run("NoLogs", func(t *testing.T) {
		// Test case when no logs exist
		// All connection counts should be 0
		param := src.CountConnectionsParam{
			EndTime: 3,
			Period:  1,
			Logs:    []src.Log{},
		}
		result := src.CountConnections(param)
		expected := []int{0, 0, 0, 0}
		assert.Equal(t, expected, result)
	})

	t.Run("GapLogs", func(t *testing.T) {
		// Test case when there are gaps between logs
		// end_time=6, period=2 with multiple logs to aggregate connection counts
		// Expected result: [0, 4, 4, 5]
		param := src.CountConnectionsParam{
			EndTime: 6,
			Period:  2,
			Logs: []src.Log{
				{Time: 1, NConnect: 4, NDisconnect: 0},
				{Time: 3, NConnect: 1, NDisconnect: 1},
				{Time: 6, NConnect: 3, NDisconnect: 2},
			},
		}
		result := src.CountConnections(param)
		expected := []int{0, 4, 4, 5}
		assert.Equal(t, expected, result)
	})

	t.Run("SingleLog", func(t *testing.T) {
		// Test case when only one log exists
		// end_time=5, period=1 with only one log to aggregate connection counts
		// Expected result: [5, 5, 5, 5, 5, 5]
		param := src.CountConnectionsParam{
			EndTime: 5,
			Period:  1,
			Logs: []src.Log{
				{Time: 0, NConnect: 5, NDisconnect: 0},
			},
		}
		result := src.CountConnections(param)
		expected := []int{5, 5, 5, 5, 5, 5}
		assert.Equal(t, expected, result)
	})

	t.Run("AllDisconnect", func(t *testing.T) {
		// Test case when all connections are disconnected
		// All connections are disconnected at time 2, expected result: [10, 10, 0, 0, 0]
		param := src.CountConnectionsParam{
			EndTime: 4,
			Period:  1,
			Logs: []src.Log{
				{Time: 0, NConnect: 10, NDisconnect: 0},
				{Time: 2, NConnect: 0, NDisconnect: 10},
			},
		}
		result := src.CountConnections(param)
		expected := []int{10, 10, 0, 0, 0}
		assert.Equal(t, expected, result)
	})

	t.Run("PartialDisconnect", func(t *testing.T) {
		// Test case when partial disconnections occur
		// end_time=4, period=1 with partial disconnections
		// Expected result: [10, 10, 5, 5, 5]
		param := src.CountConnectionsParam{
			EndTime: 4,
			Period:  1,
			Logs: []src.Log{
				{Time: 0, NConnect: 10, NDisconnect: 0},
				{Time: 2, NConnect: 0, NDisconnect: 5},
			},
		}
		result := src.CountConnections(param)
		expected := []int{10, 10, 5, 5, 5}
		assert.Equal(t, expected, result)
	})

	t.Run("LargeConnectAndDisconnect", func(t *testing.T) {
		// Test case with very large number of connections followed by full disconnection
		// end_time=100, period=10 with massive connections followed by full disconnection
		// Expected result: [0, 10000000000, 10000000000, 10000000000, 0, 0, 0, 0, 0, 0, 0]
		largeNum := int(1e10)
		param := src.CountConnectionsParam{
			EndTime: 100,
			Period:  10,
			Logs: []src.Log{
				{Time: 10, NConnect: largeNum, NDisconnect: 0},
				{Time: 40, NConnect: 0, NDisconnect: largeNum},
			},
		}
		result := src.CountConnections(param)
		expected := []int{0, largeNum, largeNum, largeNum, 0, 0, 0, 0, 0, 0, 0}
		assert.Equal(t, expected, result)
	})

	t.Run("Output1000Entries", func(t *testing.T) {
		// Boundary test
		// end_time=1000, period=1 with output for each time interval
		// Verify that connection counts are correctly calculated for all times
		var logs []src.Log
		for i := 0; i <= 1000; i++ {
			logs = append(logs, src.Log{Time: i, NConnect: 1, NDisconnect: 0})
		}
		param := src.CountConnectionsParam{
			EndTime: 1000,
			Period:  1,
			Logs:    logs,
		}
		result := src.CountConnections(param)

		var expected []int
		for i := 0; i <= 1000; i++ {
			expected = append(expected, i+1)
		}
		assert.Equal(t, expected, result)
	})

	t.Run("BoundaryPeriodEqualsEndTime", func(t *testing.T) {
		// Test case when period equals end_time
		// end_time=10, period=10, expected result: [10, 10]
		param := src.CountConnectionsParam{
			EndTime: 10,
			Period:  10,
			Logs: []src.Log{
				{Time: 0, NConnect: 10, NDisconnect: 0},
			},
		}
		result := src.CountConnections(param)
		expected := []int{10, 10}
		assert.Equal(t, expected, result)
	})

	t.Run("OverlappingConnectionsAndDisconnections", func(t *testing.T) {
		// Test case when connections and disconnections overlap
		// Test situation where connections and disconnections occur at the same time
		param := src.CountConnectionsParam{
			EndTime: 5,
			Period:  1,
			Logs: []src.Log{
				{Time: 0, NConnect: 5, NDisconnect: 0},
				{Time: 2, NConnect: 3, NDisconnect: 2}, // 3 connect, 2 disconnect at time 2
				{Time: 4, NConnect: 0, NDisconnect: 6}, // All disconnect at time 4
			},
		}
		result := src.CountConnections(param)
		expected := []int{5, 5, 6, 6, 0, 0}
		assert.Equal(t, expected, result)
	})

	t.Run("EdgeCaseZeroEndTime", func(t *testing.T) {
		// Edge case: end_time=0
		param := src.CountConnectionsParam{
			EndTime: 0,
			Period:  1,
			Logs: []src.Log{
				{Time: 0, NConnect: 5, NDisconnect: 0},
			},
		}
		result := src.CountConnections(param)
		expected := []int{5}
		assert.Equal(t, expected, result)
	})

	t.Run("LogsOutsideTimeRange", func(t *testing.T) {
		// Test case with logs outside the time range
		param := src.CountConnectionsParam{
			EndTime: 5,
			Period:  1,
			Logs: []src.Log{
				{Time: 0, NConnect: 3, NDisconnect: 0},
				{Time: 10, NConnect: 5, NDisconnect: 0}, // This should be ignored
				{Time: 2, NConnect: 2, NDisconnect: 1},
			},
		}
		result := src.CountConnections(param)
		expected := []int{3, 3, 4, 4, 4, 4}
		assert.Equal(t, expected, result)
	})
}
